lint/long-chain_checker.go

```diff
@@ -0,0 +1,121 @@
+package lint
+
+import (
+	"fmt"
+	"go/ast"
+	"strings"
+
+	"github.com/Quasilyte/astcmp"
+)
+
+type longChainChecker struct {
+	ctx *Context
+}
+
+func newLongChainChecker(ctx *Context) Checker {
+	return &longChainChecker{
+		ctx: ctx,
+	}
+}
+
+func (c *longChainChecker) Check(f *ast.File) {
+	for _, decl := range f.Decls {
+		if decl, ok := decl.(*ast.FuncDecl); ok {
+			if decl.Body == nil {
+				continue
+			}
+			for _, stmt := range decl.Body.List {
+				if stmt, ok := stmt.(*ast.SwitchStmt); ok {
+					c.checkSwitch(stmt)
+				}
+			}
+		}
+	}
+}
+
+func (c *longChainChecker) checkSwitch(stmt *ast.SwitchStmt) {
+	exprs := []ast.Expr{}
+	for _, st := range stmt.Body.List {
+		s := st.(*ast.CaseClause)
+		if s.List == nil {
+			continue
+		}
+		exprs = append(exprs, s.List...)
+	}
+	if len(exprs) < 2 {
+		return
+	}
+
+	cp := c.exprToList(exprs[0])
+	for _, e := range exprs[1:] {
+		cp = c.commonPrefix(cp, c.exprToList(e))
+	}
+
+	const n = 3
+
+	if len(cp) > n {
+		c.warn(cp, stmt)
+	}
+}
+
+func (c *longChainChecker) exprToList(expr ast.Expr) []ast.Expr {
+	res := []ast.Expr{}
+	tmp := expr
+	for {
+		switch t := tmp.(type) {
+		case *ast.SelectorExpr:
+			res = append(res, t.Sel)
+			tmp = t.X
+		default:
+			res = append(res, tmp)
+			res = reverseExprSlice(res)
+			return res
+
+		}
+	}
+}
+
+func reverseExprSlice(e []ast.Expr) []ast.Expr {
+	for i := len(e)/2 - 1; i >= 0; i-- {
+		j := len(e) - 1 - i
+		e[i], e[j] = e[j], e[i]
+	}
+	return e
+}
+
+func (c *longChainChecker) commonPrefix(a, b []ast.Expr) []ast.Expr {
+	res := []ast.Expr{}
+
+	len := c.min(len(a), len(b))
+
+	for i := 0; i < len; i++ {
+		if astcmp.EqualExpr(a[i], b[i]) {
+			res = append(res, a[i])
+		} else {
+			break
+		}
+	}
+
+	return res
+}
+
+func (c *longChainChecker) min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func (c *longChainChecker) warn(exprs []ast.Expr, stmt ast.Stmt) {
+	var s []string
+	for _, e := range exprs {
+		s = append(s, nodeString(c.ctx.FileSet, e))
+	}
```

Comment:
Not very performance-critical application and method, but here is advice that could be useful in future:
when you know exact length of slice you want to fill, it's possible to reserve the memory once.
```diff
+	var s []string
+	for _, e := range exprs {
+		s = append(s, nodeString(c.ctx.FileSet, e))
+	}
-	s := make([]string, len(exprs))
-	for i, e := range exprs {
-		s[i] = nodeString(c.ctx.FileSet, e)
-	}
```
Also note that you can avoid append calls with direct index assignment.
If you want to keep append there, `len(exprs)` should be used as a `cap` argument, not `len` argument:
```go

s := make([]string, 0, len(exprs))
for _, e := range exprs {
	s = append(s, nodeString(c.ctx.FileSet, e))
}
```
Note the `0` length.

Reply:
Good advice, thx

Reply:
fixed